content/precise-types.md

```diff
@@ -0,0 +1,517 @@
+val v1T: Vec[1] = Vec[1](1)
+val v2T: Vec[2] = Vec[2](2)
+val v3 = v1T ++ v2T
+val v3T: Vec[3] = v3 // error
+~~~
+
+*Output*:
+~~~ scala
+-- [E007] Type Mismatch Error: vec.scala:16:18 -------------------
+16 |val v3T: Vec[3] = v3 // error
+   |                  ^^                                         
+   |                  Found:    (v3 : Vec[Int + Int])
+   |                  Required: Vec[(3 : Int)]
+~~~
+Implicit summoning of a covariant type parameter causes widening. To avoid widening here we can define the type class `TC` as invariant (and modify the related given declarations), or drop the covariance from `Vec`, or add a `Singleton` upper-bound to `S0`.
+
+As seen from all the examples thus far, the scala compiler is very eager to widen types in various cases and we need a mechanism to prevent this when the compile-time information preservation is crucial.
+
+## Proposed solution
+
+### High-level overview
+
+We propose to add a `precise` modifier that can be applied on type parameters to indicate to the compiler that term arguments annotated with those precise types are considered to be *precisely typed*. Precise typing is not only preventing widening for literals as demonstrated so far, but also prevents any widening in the expression composition, so that composable expressions like tuples can preserve their precise representation throughout their entire structure.
+
+Here is our earlier implementation of `Vec`, modified to work with all examples by applying the `precise` modifier:
+~~~ scala
+class Vec[precise +S <: Int](val size: S):
+  def ++[precise TS <: Int](that: Vec[TS]): Vec[S + TS] = 
+    Vec[S + TS]((size + that.size).asInstanceOf[S + TS])
+~~~
+The `precise` modifier helps preserving the precise size type when constructing a vector (precise typing of the size term argument), and when the vector is applied as an argument to the concatenation (not widening the covariant size type parameter).
+
+
+### Specification
+
+#### Add The `precise` Modifier
+The precise modifier is a soft keyword that is only available at a type parameter declaration position.
+The following grammar rules are modified to support this new soft keyword:
+~~~
+ClsTypeParam ::=  {Annotation} [‘precise’] [‘+’ | ‘-’] id [HkTypeParamClause] TypeParamBounds ;
+DefTypeParam ::=  {Annotation} [‘precise’] id [HkTypeParamClause] TypeParamBounds ;
+TypTypeParam ::=  {Annotation} [‘precise’] id [HkTypeParamClause] TypeBounds ;
+HkTypeParam  ::=  {Annotation} [‘precise’] [‘+’ | ‘-’] (id [HkTypeParamClause] | ‘_’) TypeBounds ;
+~~~ 
+
+#### Precise Types
+A type is considered to be precise when at least of one of following occurs:
+  * It is a type parameter declaration that is directly modified with `precise` 
+  ~~~ scala
+  //`T` is a precise type parameter declaration
+  class PreciseBox[precise T]
+  ~~~
+  * It is a type variable reference with a precise type parameter origin
```

Comment:
This would need to be defined in more detail. What if a type variable appears in a union or intersection of an instance? If it is an alias of an instance? One would need to enumerate all possibilities as so far the concept of "origin" is not defined in Scala type inference.

Reply:
> This would need to be defined in more detail. What if a type variable appears in a union or intersection of an instance? If it is an alias of an instance? One would need to enumerate all possibilities as so far the concept of "origin" is not defined in Scala type inference.

All the cases that are considered are explicitly specified, so union/intersection/aliases of precise type parameters are not considered. I'll gladly change "origin" to the proper terminology. Any suggestion to modify the spec will be appreciated.